### PR TITLE
Add support for implementation edges in our LSIF indexer

### PR DESCRIPTION
**Pull request by jasonmalinowski**
Fixes https://github.com/dotnet/roslyn/issues/59617

### CODE REVIEW

File: src/Features/Lsif/Generator/Generator.cs [hunk 1]
```diff
@@ -228,6 +229,29 @@ SymbolKind.RangeVariable or
...
+                        // implement the interface, for example in this case:
+                        //
+                        //     interface I { void M(); }
+                        //     class Base { public void M() { } }
+                        //     class Derived : Base, I { }
+                        //
+                        // We don't worry about supporting these cases here.
+                        var implementedMembers = declaredSymbol.ExplicitOrImplicitInterfaceImplementations();
+
+                        foreach (var implementedMember in implementedMembers)
```
Comment (jasonmalinowski) [inline src/Features/Lsif/Generator/Generator.cs:244]: We're still discussing exactly how override methods should be linked, so that'll be a separate PR.

File: src/Features/Lsif/Generator/Generator.cs [hunk 2]
```diff
@@ -251,6 +251,45 @@ SymbolKind.RangeVariable or
...
+                            var overridenMember = declaredSymbol.GetOverriddenMember();
+
+                            while (overridenMember?.GetOverriddenMember() != null)
+                                overridenMember = overridenMember.GetOverriddenMember();
+
+                            if (overridenMember != null)
+                                MarkImplementationOfSymbol(overridenMember);
+                        }
+
+                        void MarkImplementationOfSymbol(ISymbol baseMember)
```
Comment (jasonmalinowski) [inline src/Features/Lsif/Generator/Generator.cs:282]: The only way this will make sense is if you look at the picture in https://github.com/dotnet/roslyn/issues/59617.

File: src/Features/Lsif/Generator/Generator.cs [hunk 3]
```diff
@@ -251,6 +251,45 @@ SymbolKind.RangeVariable or
...
+                        //     class Derived : Base, I { }
+                        //
+                        // We don't worry about supporting these cases here.
+                        var implementedMembers = declaredSymbol.ExplicitOrImplicitInterfaceImplementations();
+
+                        foreach (var implementedMember in implementedMembers)
+                            MarkImplementationOfSymbol(implementedMember);
+
+                        // If this overrides a method, we'll also mark it the same way. We want to chase to the base virtual method, skipping over intermediate
+                        // methods so that way all overrides of the same method point to the same virtual method
```
Comment (dibarbet) [inline src/Features/Lsif/Generator/Generator.cs:270]: interesting - I would have thought we would have pointed to intermediate members which themselves point back to the base in a chain.  This loses the intermediate member information, but I have no idea if that is OK or not (not sure how this is used).

**Review by dibarbet — APPROVED** [review]

File: src/Features/Lsif/Generator/Generator.cs (hunk 3 above)
Comment (jasonmalinowski) [reply src/Features/Lsif/Generator/Generator.cs:270]: We discussed this and decided to go with this way for now. We saw two potential downsides of having the full chain:

1. It's not clear what happens if somebody removes their method of an override in the middle of a chain, since then the ends of the chain are disconnected.
2. It means a find references in the cloud has to connect all the chains together, as opposed to doing a faster lookup.

So we're going this way for now.
